File: packages/core/src/projection.ts
import type { ArborBlock, Materialization, MarkdownDocument, TreeChild } from "./types.ts";
import type { LogicalPath, ProtocolNodeKind, ResolvedNodeRef } from "./protocol.ts";
import { canonicalNodePath } from "./logical-path.ts";
import { relativeLogicalReference, resolveLogicalURL } from "./logical-url.ts";

/**
 * The client-derived projected directory document (spec/format.md §4).
 *
 * arbord returns storage-shaped facts — the stored or implicit body plus the
 * authoritative child listing. Clients derive this projection; it never
 * appears in a REST `NodeSnapshot.document`, and synthetic rows are never
 * serialized back through `writeMarkdown`.
 */
export interface ProjectedDocument {
  /** The stored body (or an empty document when the body is implicit). Never mutated. */
  source: MarkdownDocument;
  /** Stored blocks plus one appended synthetic row per unmatched child. */
  visibleBlocks: ArborBlock[];
  /** One entry per immediate child, in projected document order. */
  managedChildren: ManagedChildRow[];
  bodyState: "stored" | "implicit";
}

export interface ManagedChildRow {
  blockID: string;
  ref: ResolvedNodeRef;
  origin: "authored" | "synthetic";
  kind: ProtocolNodeKind;
  materialization: Materialization;
}

export interface ProjectionInput {
  path: LogicalPath;
  /** The stored document, or null when the directory body is implicit. */
  document: MarkdownDocument | null;
  /** The complete, drained child listing in authoritative server order. */
  children: readonly TreeChild[];
}

/**
 * Editor-only block ID for a synthetic managed row, derived from child
 * identity so it is stable across reprojections and sibling reorders. The
 * `managed:` prefix is load-bearing: these IDs must never be sent to the
 * server as block IDs or anchors.
 */
export function syntheticRowBlockID(child: { pageID?: string; path: string }): string {
  return child.pageID ? `managed:id:${child.pageID}` : `managed:path:${canonicalNodePath(child.path)}`;
}

export function isSyntheticRowBlockID(blockID: string): boolean {
  return blockID.startsWith("managed:");
}

/** The wire child listing still uses the fs kind vocabulary in places. */
function protocolKind(kind: TreeChild["kind"] | ProtocolNodeKind): ProtocolNodeKind {
  return kind === "postgres" ? "database" : kind;
}

const EMPTY_DOCUMENT: MarkdownDocument = {
  frontmatter: {},
  frontmatterSource: null,
  bodySource: "",
  blocks: [],
};

export function projectDirectoryDocument(input: ProjectionInput): ProjectedDocument {
  const directory = canonicalNodePath(input.path);
  const source = input.document ?? EMPTY_DOCUMENT;

  const childByPath = new Map<string, TreeChild>();
  const childByPageID = new Map<string, TreeChild>();
  for (const child of input.children) {
    childByPath.set(canonicalNodePath(child.path), child);
    if (child.pageID) childByPageID.set(child.pageID, child);
  }

  const matched = new Map<TreeChild, { blockID: string; order: number }>();
  let order = 0;

  const walk = (blocks: readonly ArborBlock[]) => {
    for (const block of blocks) {
      order += 1;
      if (block.type === "standaloneLink") {
        const link = resolveLogicalURL(directory, String(block.props?.path ?? ""));
        let child: TreeChild | undefined;
        if (link?.kind === "local") {
          // A durable ID wins over a stale path.
          child = (link.pageID && childByPageID.get(link.pageID)) || childByPath.get(link.path);
        }
        if (child && !matched.has(child)) {
          matched.set(child, { blockID: block.id, order });
        }
      }
      walk(block.children);
    }
  };
  walk(source.blocks);

  const authoredRows = [...matched.entries()]
    .sort((a, b) => a[1].order - b[1].order)
    .map(([child, { blockID }]): ManagedChildRow => ({
      blockID,
      ref: { path: canonicalNodePath(child.path), ...child.pageID ? { pageID: child.pageID } : {} },
      origin: "authored",
      kind: protocolKind(child.kind),
      materialization: child.materialization,
    }));

  const syntheticChildren = input.children.filter((child) => !matched.has(child));
  const syntheticBlocks = syntheticChildren.map((child): ArborBlock => ({
    id: syntheticRowBlockID(child),
    type: "standaloneLink",
    content: child.name,
    props: { path: relativeLogicalReference(directory, child.path) },
    children: [],
  }));
  const syntheticRows = syntheticChildren.map((child, index): ManagedChildRow => ({
    blockID: syntheticBlocks[index]!.id,
    ref: { path: canonicalNodePath(child.path), ...child.pageID ? { pageID: child.pageID } : {} },
    origin: "synthetic",
    kind: protocolKind(child.kind),
    materialization: child.materialization,
  }));

  return {
    source,
    visibleBlocks: [...source.blocks, ...syntheticBlocks],
    managedChildren: [...authoredRows, ...syntheticRows],
    bodyState: input.document ? "stored" : "implicit",
  };
}

File: native/Packages/ArborClient/Sources/ArborClient/Projection.swift
import Foundation

/// The client-derived projected directory document, mirroring
/// `@arbor/core` `projection.ts` exactly. Synthetic rows never appear in a
/// REST document and are never serialized back through `writeMarkdown`.
public struct ProjectedDocument: Codable, Sendable, Equatable {
    /// The stored body (or an empty document when the body is implicit).
    public var source: MarkdownDocument
    /// Stored blocks plus one appended synthetic row per unmatched child.
    public var visibleBlocks: [ArborBlock]
    /// One entry per immediate child, in projected document order.
    public var managedChildren: [ManagedChildRow]
    /// "stored" or "implicit".
    public var bodyState: String
}

public struct ManagedChildRow: Codable, Sendable, Equatable {
    public var blockID: String
    public var ref: ResolvedNodeRef
    /// "authored" or "synthetic".
    public var origin: String
    public var kind: String
    public var materialization: String
}

/// Editor-only block ID for a synthetic managed row, derived from child
/// identity. Byte-identical to the TypeScript derivation; the `managed:`
/// prefix must never be sent to the server.
public func syntheticRowBlockID(pageID: String?, path: String) -> String {
    if let pageID { return "managed:id:\(pageID)" }
    return "managed:path:\(canonicalNodePath(path))"
}

public func isSyntheticRowBlockID(_ blockID: String) -> Bool {
    blockID.hasPrefix("managed:")
}

private func protocolKind(_ kind: String) -> String {
    kind == "postgres" ? "database" : kind
}

public func projectDirectoryDocument(
    path inputPath: String,
    document: MarkdownDocument?,
    children: [TreeChild]
) -> ProjectedDocument {
    let directory = canonicalNodePath(inputPath)
    let source = document ?? MarkdownDocument(frontmatter: [:], frontmatterSource: nil, bodySource: "", blocks: [])

    var childByPath: [String: Int] = [:]
    var childByPageID: [String: Int] = [:]
    for (index, child) in children.enumerated() {
        childByPath[canonicalNodePath(child.path)] = index
        if let pageID = child.pageID { childByPageID[pageID] = index }
    }

    var matched: [Int: (blockID: String, order: Int)] = [:]
    var order = 0

    func walk(_ blocks: [ArborBlock]) {
        for block in blocks {
            order += 1
            if block.type == "standaloneLink" {
                var href = ""
                if case .string(let value)? = block.props?["path"] { href = value }
                if case .local(let path, let pageID, _)? = resolveLogicalURL(base: directory, href: href) {
                    // A durable ID wins over a stale path.
                    let index = pageID.flatMap { childByPageID[$0] } ?? childByPath[path]
                    if let index, matched[index] == nil {
                        matched[index] = (block.id, order)
                    }
                }
            }
            walk(block.children)
        }
    }
    walk(source.blocks)

    let authoredRows = matched
        .sorted { $0.value.order < $1.value.order }
        .map { entry -> ManagedChildRow in
            let child = children[entry.key]
            return ManagedChildRow(
                blockID: entry.value.blockID,
                ref: ResolvedNodeRef(path: canonicalNodePath(child.path), pageID: child.pageID),
                origin: "authored",
                kind: protocolKind(child.kind),
                materialization: child.materialization
            )
        }

    let syntheticChildren = children.enumerated().filter { matched[$0.offset] == nil }.map(\.element)
    let syntheticBlocks = syntheticChildren.map { child in
        ArborBlock(
            id: syntheticRowBlockID(pageID: child.pageID, path: child.path),
            type: "standaloneLink",
            content: child.name,
            props: ["path": .string(relativeLogicalReference(from: directory, to: child.path))]
        )
    }
    let syntheticRows = zip(syntheticChildren, syntheticBlocks).map { child, block in
        ManagedChildRow(
            blockID: block.id,
            ref: ResolvedNodeRef(path: canonicalNodePath(child.path), pageID: child.pageID),
            origin: "synthetic",
            kind: protocolKind(child.kind),
            materialization: child.materialization
        )
    }

    return ProjectedDocument(
        source: source,
        visibleBlocks: source.blocks + syntheticBlocks,
        managedChildren: authoredRows + syntheticRows,
        bodyState: document == nil ? "implicit" : "stored"
    )
}

public enum ProjectedNodeUpdate: Sendable {
    case event(WorkspaceEvent)
    case resync(NodeSnapshot, ProjectedDocument?)
}

public struct ProjectedNodeView: Sendable {
    /// The hydrated raw snapshot. `document` never contains synthetic rows.
    public var snapshot: NodeSnapshot
    /// The derived directory projection; nil for non-directory nodes.
    public var projection: ProjectedDocument?
    public var updates: AsyncThrowingStream<ProjectedNodeUpdate, Error>
}

/// Derive the projection from a hydrated snapshot; nil for nodes without a
/// child-bearing surface. Mirrors the TypeScript `projectSnapshot`.
public func projectSnapshot(_ snapshot: NodeSnapshot) -> ProjectedDocument? {
    guard snapshot.kind == "directory" || snapshot.kind == "collection" else { return nil }
    return projectDirectoryDocument(
        path: snapshot.path,
        document: snapshot.bodyState == "implicit" ? nil : snapshot.document,
        children: snapshot.children ?? []
    )
}
